File: benchmarks/fannkuchredux.java
/* The Computer Language Benchmarks Game
   http://shootout.alioth.debian.org/

   contributed by Isaac Gouy
   converted to Java by Oleg Mazurov
*/

public class fannkuchredux
{
   public static int fannkuch(int n) {
      int[] perm = new int[n];
      int[] perm1 = new int[n];
      int[] count = new int[n];
      int maxFlipsCount = 0;
      int permCount = 0;
      int checksum = 0;

      for(int i=0; i<n; i++) perm1[i] = i;
      int r = n;

      while (true) {

         while (r != 1){ count[r-1] = r; r--; }

         for(int i=0; i<n; i++) perm[i] = perm1[i];
         int flipsCount = 0;
         int k;

         while ( !((k=perm[0]) == 0) ) {
            int k2 = (k+1) >> 1;
            for(int i=0; i<k2; i++) {
               int temp = perm[i]; perm[i] = perm[k-i]; perm[k-i] = temp;
            }
            flipsCount++;
         }

         maxFlipsCount = Math.max(maxFlipsCount, flipsCount);
         checksum += permCount%2 == 0 ? flipsCount : -flipsCount;

         // Use incremental change to generate another permutation
         while (true) {
            if (r == n) {
	             return checksum;
	          }
            int perm0 = perm1[0];
            int i = 0;
            while (i < r) {
               int j = i + 1;
               perm1[i] = perm1[j];
               i = j;
            }
            perm1[r] = perm0;

            count[r] = count[r] - 1;
            if (count[r] > 0) break;
            r++;
         }

         permCount++;
      }
   }

   public static void main(String[] args){
      int numIterations = Integer.valueOf(args[0]);
      int warmUp        = Integer.valueOf(args[1]);
      int problemSize   = Integer.valueOf(args[2]);

      if (fannkuch(9) != 8629) {
        System.err.println("fannkuch(9) failed, gave unexpected result.");
        System.exit(1);
      }

      for (int i = 0; i < warmUp; i++) {
        fannkuch(problemSize);
      }

      int result = 0;
      for (int i = 0; i < numIterations; i++) {
        long start = System.nanoTime();
        result += fannkuch(problemSize);
        long end = System.nanoTime();
        long microseconds = (end - start) / 1000;
        System.out.println("Fannkuch: iterations=1 runtime: " + microseconds + "us");
      }
   }
}

File: benchmarks/fannkuchredux.py
#    The Computer Language Benchmarks Game
#    http://shootout.alioth.debian.org/

#    contributed by Isaac Gouy
#    converted to Java by Oleg Mazurov
#    converted to Python by Buck Golemon
#    modified by Justin Peel

def fannkuch(n):
    maxFlipsCount = 0
    permSign = True
    checksum = 0

    perm1 = list(range(n))
    count = perm1[:]
    rxrange = range(2, n - 1)
    nm = n - 1
    while 1:
        k = perm1[0]
        if k:
            perm = perm1[:]
            flipsCount = 1
            kk = perm[k]
            while kk:
                perm[:k+1] = perm[k::-1]
                flipsCount += 1
                k = kk
                kk = perm[kk]
            if maxFlipsCount < flipsCount:
                maxFlipsCount = flipsCount
            checksum += flipsCount if permSign else -flipsCount

        # Use incremental change to generate another permutation
        if permSign:
            perm1[0],perm1[1] = perm1[1],perm1[0]
            permSign = False
        else:
            perm1[1],perm1[2] = perm1[2],perm1[1]
            permSign = True
            for r in rxrange:
                if count[r]:
                    break
                count[r] = r
                perm0 = perm1[0]
                perm1[:r+1] = perm1[1:r+2]
                perm1[r+1] = perm0
            else:
                r = nm
                if not count[r]:
                    return checksum
            count[r] -= 1

import sys, time

def microseconds():
    return time.time() * 1000 * 1000



iterations   = 100
warmup       = 0
problem_size = 1000

if len(sys.argv) > 1:
    iterations = int(sys.argv[1])

if len(sys.argv) > 2:
    warmup = int(sys.argv[2])

if len(sys.argv) > 3:
    problem_size = int(sys.argv[3])

print "Overall iterations: %d." % iterations
print "Warmup  iterations: %d." % warmup
print "Problem size:       %d." % problem_size

for i in xrange(0, warmup):
    fannkuch(problem_size)

if fannkuch(9) != 8629:
    print "Sanity check failed! Fannkuch gives wrong result"
    sys.exit(1)

for i in xrange(0, iterations):
    start = microseconds()
    fannkuch(problem_size)
    elapsed = microseconds() - start
    print "Fannkuch: iterations=1 runtime: %dus" % elapsed
